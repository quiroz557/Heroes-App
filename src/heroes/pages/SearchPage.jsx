import React from 'react'
import { HeroCard } from '../components/HeroCard'
import { useForm } from '../../hooks'
import { useLocation, useNavigate } from 'react-router-dom'
import queryString from 'query-string'
import { getHeroesByName } from '../helpers'

export const SearchPage = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse( location.search );
  const heroes = getHeroesByName(q);

  const showSearch = (q.length === 0);
  const showError = ( heroes.length === 0 && q.length > 0 );

  const { searchHero, onInputChange } = useForm({
    searchHero: q
  });

  const onSearchSubmit = ( event ) => {
    event.preventDefault();

    // if ( searchHero.trim().length <= 1 ) return;

    navigate(`?q=${ searchHero }`)
  }

  return (
    <>
       <h1>SearchPage</h1>

       <div className="row">

        <div className="col-5">
            <h4>Searching</h4>
            <hr />
            <form aria-label='form' onSubmit={ onSearchSubmit }>
              <input 
                type={'text'}
                placeholder={'Search a hero'}
                className='form-control'
                name='searchHero'
                autoComplete='off'
                onChange={onInputChange}
                value={searchHero}
              />

              <button className='btn btn-outline-primary mt-1'>
                Search
              </button>

            </form>
        </div>

        <div className="col-7">
            <h4>Results</h4>
            <hr />

            <div 
              className="alert alert-primary" 
              style={{display: showSearch ? '' : 'none'}}
              aria-label='div-search'
            >
              Search a hero
            </div>

            <div  
              aria-label='div-error'
              className="alert alert-danger" 
              style={{display: showError ? '' : 'none'}}
            >
              No hero with <b>{ q }</b>
            </div>

            {
              heroes.map( hero =>  (
                <HeroCard key={hero.id} {...hero} />
              ))
            }
            
        </div>

       </div>

    </>
  )
}
