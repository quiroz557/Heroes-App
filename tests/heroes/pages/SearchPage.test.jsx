import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { SearchPage } from "../../../src/heroes/pages/SearchPage"

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate,
}));

describe('Pruebas en <SearchPage />', () => { 

    beforeEach(() => jest.clearAllMocks());

    test('Debe de mostrarse correctamente con valores por defecto', () => { 
        
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        expect( container ).toMatchSnapshot();

    });

    test('Debe de mostrar a Batman y el input con el valor del queryString', () => { 
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        );

       const input = screen.getByRole('textbox');
       expect( input.value ).toBe('batman');

       const img = screen.getByRole('img');
       expect( img.src ).toContain('/src/assets/heroes/dc-batman.jpg');

       const divSearch = screen.getByLabelText('div-search');
       expect( divSearch.style.display ).toBe('none');
        
    });

    test('Debe de mostrar un error si no se encuentra el hero', () => {  
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        );

        const divError = screen.getByLabelText('div-error');
        expect( divError.style.display ).toBe('');
    })

    test('Debe de llamar el navigate a la pantalla nueva', () => { 
        
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change(input, {target: {value: 'batman'}})

        const form = screen.getByLabelText('form');
        fireEvent.submit(form);

        expect( mockedUsedNavigate ).toHaveBeenCalledWith('?q=batman');
    });
});