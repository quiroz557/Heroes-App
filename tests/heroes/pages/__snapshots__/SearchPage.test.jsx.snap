// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchPage
  </h1>
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchHero"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="div-search"
        class="alert alert-primary"
      >
        Search a hero
      </div>
      <div
        aria-label="div-error"
        class="alert alert-danger"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
